hierarchicalforecast/methods.py

```diff
@@ -59,23 +64,243 @@ def _reconcile(S: np.ndarray, P: np.ndarray, W: np.ndarray,
+        self.y_hat_insample = y_hat_insample
+        self.y_hat = y_hat
+        self.n_samples = n_samples
+        self.seed = seed
+    
+    def get_samples(self):
+        residuals = self.y_insample - self.y_hat_insample
+        h = self.y_hat.shape[1]
+        #removing nas from residuals
+        residuals = residuals[:, np.isnan(residuals).sum(axis=0) == 0]
+        sample_idx = np.arange(residuals.shape[1] - h)
+        state = np.random.RandomState(self.seed)
+        samples_idx = state.choice(sample_idx, size=self.n_samples)
+        samples = [self.y_hat + residuals[:, idx:(idx + h)] for idx in samples_idx]
+        return np.stack(samples)
 
 # %% ../nbs/methods.ipynb 7
+class PERMBU:
+    """PERMBU Probabilistic Reconciliation Class.
+
+    The PERM-BU method leverages empirical bottom-level marginal distributions 
+    with empirical copula functions (describing bottom-level dependencies) to 
+    generate the distribution of aggregate-level distributions using BottomUp 
+    reconciliation. The sample reordering technique in the PERM-BU method reinjects 
+    multivariate dependencies into independent bottom-level samples.
+
+    **References:**<br>
+    - [Taieb, Souhaib Ben and Taylor, James W and Hyndman, Rob J. (2017). 
+    Coherent probabilistic forecasts for hierarchical time series. 
+    International conference on machine learning ICML.](https://proceedings.mlr.press/v70/taieb17a.html)
+    """
+    def __init__(
+            self,
+            y_insample: np.ndarray,
+            y_hat_insample: np.ndarray,
+            sigmah: np.ndarray,
+            S: np.ndarray,                 
+            n_samples: int=None,
+            seed: int=0
+        ):
+        """
+        **Parameters:**<br>
+        `S`: Summing matrix of size (`base`, `bottom`).<br>
+        `sigmah`: Forecast standard dev. of size (`base`, `horizon`).<br>
+        `y_insample`: Insample values of size (`base`, `insample_size`).<br>
+        `y_hat_insample`: Insample values of size (`base`, `insample_size`).<br>
+        `n_samples`: int, number of normal prediction samples generated.<br>
+        """
+        self.y_insample = y_insample
+        self.y_hat_insample = y_hat_insample
+        self.sigmah = sigmah
+        self.S = S
+        self.n_samples = n_samples
+        self.seed = seed
+    
+    def _obtain_ranks(self, array):
+        """ Vector ranks
+
+        Efficiently obtain vector ranks.
+        Example `array=[4,2,7,1]` -> `ranks=[2, 1, 3, 0]`.
+
+        **Parameters**<br>
+        `array`: np.array, matrix with floats or integers on which the 
+                ranks will be computed on the second dimension.<br>
+
+        **Returns**<br>
+        `ranks`: np.array, matrix with ranks along the second dimension.<br>
+        """
+        temp = array.argsort(axis=1)
+        ranks = np.empty_like(temp)
+        a_range = np.arange(temp.shape[1])
+        for iRow in range(temp.shape[0]):
```

Comment:
I would keep caps for classes and underscores for other variables i_row?

Reply:
Sure. I can change it. To be completely transparent, I didn't change anything in the PERMBU class. I just made it compatible. The `iRow` variable was merged into the main branch without review.